[OMPIRBuilder] Don't generate DISubprogram for outlined function.

abidh opened the pull request:
In `OMPIRBuilder`, a new function is created for the `TargetOp`. We also create a new `DISubprogram` for it. All the variables that were in the target region now have to be updated to have the correct scope. This after the fact updating of debug information becomes very difficult in certain cases.

This PR is making the change that `OMPIRBuilder` will not generate a `DISubprogram`. The responsibility has been shifted to the frontend. This allows us to simplify the updating of debug records.

The PR is made a bit more complicated by the the fact that in new scheme, the debug location already points to the new `DISubprogram` by the time it reaches `convertOmpTarget`. But we need some code generation in the parent function so we have to carefully manage the debug locations.

This builds on https://github.com/llvm/llvm-project/pull/138039 and fixes issue https://github.com/llvm/llvm-project/issues/134991.

abidh committed:
[OpenMP] Improve debug info generation for outlined function.

In OMPIRBuilder, a new function is created for the TargetOp. We also
create a new DISubprogram for it. All the variables that were in the
target region now have to be updated to have the correct scope. This
after the fact updating of debug information becomes very difficult in
certain cases.

This PR is making the change that OMPIRBuilder will not generate a
DISubprogram. The responsibility has been shifted to the frontend. This
allows us to simplify the updating of debug records.

The PR is made a bit more complicated by the the fact that in new
scheme, the debug location already points to the new DISubprogram by
the time it reaches convertOmpTarget. But we need some code generation
in the parent function so we have to carefully manage the debug
locations.

This builds on `#138039` and fixes issue `#134991`.

abidh commented:
Polite ping.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -4982,6 +4982,20 @@ static LogicalResult
+  // correct debug locations. For outlined functions, the normal MLIR op
+  // conversion will automatically pick the correct location.
+  llvm::BasicBlock *parentBB = builder.GetInsertBlock();
+  if (parentBB && !parentBB->empty())
+    builder.SetCurrentDebugLocation(parentBB->back().getDebugLoc());
+  else
+    builder.SetCurrentDebugLocation(llvm::DebugLoc());
```
If I understand it correctly, this will use the debug information from the last instruction introduced in this block (i.e. the last thing added while translating the previous MLIR op). But then it will not do so if `omp.target` is the first operation in the block, which doesn't seem right.

Looking at the comment above, would it make sense to perhaps get the debug information from `parentBB->getParent()` (the parent function)? Also, couldn't we assert that `parentBB != nullptr` here?

abidh replied:
My idea was that for empty basic block, an empty location will be set. But it occurred to me that it will cause Verifier failure in certain cases. I have now changed the way debug location is set and added a testcase to cover it.

skatrak reviewed (commented):
Thanks Abid! Just a small comment from me trying to understand how this works.

abidh committed:
Handle review comments.

The new way to set debug location can better handle empty basic block or function. A testcase is added as well.

abidh commented:
Polite ping.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -4982,9 +4982,27 @@ static LogicalResult
+  // get the correct debug locations. For outlined functions, the normal MLIR op
+  // conversion will automatically pick the correct location.
+  llvm::BasicBlock *parentBB = builder.GetInsertBlock();
+  assert(parentBB);
```
Nit: Add some minimally descriptive message to this assert and the one below.

abidh replied:
Done.

skatrak reviewed (approved):
Thank you again Abid, LGTM. Apologies for the delay taking another look at this!

abidh committed:
Add message in asserts.
